File: src/main/java/io/github/slimefunguguproject/bump/core/services/sounds/SoundService.java
package io.github.slimefunguguproject.bump.core.services.sounds;

import java.util.EnumMap;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Level;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.annotation.ParametersAreNonnullByDefault;

import com.google.common.base.Preconditions;

import io.github.slimefunguguproject.bump.implementation.Bump;

import net.guizhanss.guizhanlib.slimefun.addon.AddonConfig;

/**
 * This service manages all sounds used in Bump.
 *
 * @author ybw0014
 */
public final class SoundService {

    private final AddonConfig config;

    private final Map<BumpSound, SoundConfig> soundMap = new EnumMap<>(BumpSound.class);

    /**
     * Initialize {@link SoundService}.
     *
     * @param soundConfig The config file of sounds.
     */
    public SoundService(@Nonnull AddonConfig soundConfig) {
        this.config = soundConfig;
    }

    /**
     * Load the configurations of all {@link BumpSound}.
     *
     * @param saveConfig Whether to save config after loading.
     */
    public void load(boolean saveConfig) {
        config.reload();

        for (BumpSound sound : BumpSound.values()) {
            try {
                loadSound(sound);
            } catch (Exception ex) {
                Bump.getLocalization().log(Level.SEVERE, "error-sound", sound.name());
            }
        }

        if (saveConfig) {
            config.save();
        }
    }

    /**
     * Get the {@link SoundConfig} for a specified {@link BumpSound}.
     *
     * @param sound The {@link BumpSound} to get {@link SoundConfig} for.
     *
     * @return The {@link SoundConfig} for the {@link BumpSound}. {@code null} if not exists.
     */
    @Nullable
    public SoundConfig getSoundConfig(@Nonnull BumpSound sound) {
        Preconditions.checkArgument(sound != null, "BumpSound cannot be null");

        return soundMap.get(sound);
    }

    private void loadSound(@Nonnull BumpSound sound) {
        setDefault(sound.name() + ".sound", sound.getSound());
        setDefault(sound.name() + ".volume", sound.getVolume());
        setDefault(sound.name() + ".pitch", sound.getPitch());

        String soundId = config.getString(sound.name() + ".sound");
        float volume = (float) config.getDouble(sound.name() + ".volume");
        float pitch = (float) config.getDouble(sound.name() + ".pitch");

        if (volume < 0) {
            Bump.getLocalization().log(Level.WARNING, "invalid-sound-volume", sound.name(), volume, 0.0F);
            volume = 0;
        }

        if (pitch < 0.5F) {
            Bump.getLocalization().log(Level.WARNING, "invalid-sound-pitch", sound.name(), volume, 0.5F);
            pitch = 0.5F;
        }

        SoundConfig soundConfig = new SoundConfig(soundId, volume, pitch);
        soundMap.put(sound, soundConfig);
    }

    @ParametersAreNonnullByDefault
    private void setDefault(String path, Object value) {
        if (!config.contains(path)) {
            config.set(path, value);
        }
    }
}

File: src/main/kotlin/io/github/slimefunguguproject/bump/core/services/sounds/SoundService.kt
package io.github.slimefunguguproject.bump.core.services.sounds

import io.github.slimefunguguproject.bump.Bump
import io.github.slimefunguguproject.bump.utils.GeneralUtils.valueOfOrNull
import net.guizhanss.guizhanlib.slimefun.addon.AddonConfig
import org.bukkit.Sound
import java.util.logging.Level
import kotlin.math.max

class SoundService(private val config: AddonConfig) {
    private val soundMap: MutableMap<BumpSound, SoundConfig> = mutableMapOf()

    /**
     * Load the configurations of all [BumpSound].
     *
     * @param saveConfig Whether to save config after loading.
     */
    fun load(saveConfig: Boolean) {
        config.reload()

        BumpSound.entries.forEach {
            try {
                loadSound(it)
            } catch (ex: Exception) {
                Bump.log(Level.SEVERE, ex, "An error occurred while loading sound config for ${it.name}.")
            }
        }

        if (saveConfig) {
            config.save()
        }
    }

    /**
     * Get the [SoundConfig] for a specified [BumpSound].
     *
     * @param sound The [BumpSound] to get [SoundConfig] for.
     *
     * @return The [SoundConfig] for the [BumpSound]. `null` if not exists.
     */
    internal fun getSoundConfig(sound: BumpSound): SoundConfig? {
        return soundMap[sound]
    }

    private fun loadSound(sound: BumpSound) {
        setDefault(sound.name + ".sound", sound.sound.name)
        setDefault(sound.name + ".volume", sound.volume)
        setDefault(sound.name + ".pitch", sound.pitch)

        val soundId = config.getString(sound.name + ".sound") ?: sound.sound.name
        val soundEnum = valueOfOrNull<Sound>(soundId) ?: sound.sound
        val volume = max(config.getDouble(sound.name + ".volume"), 0.0)
        val pitch = max(config.getDouble(sound.name + ".pitch"), 0.5)

        val soundConfig = SoundConfig(soundEnum, volume.toFloat(), pitch.toFloat())
        soundMap[sound] = soundConfig
    }

    private fun setDefault(path: String, value: Any) {
        if (!config.contains(path)) {
            config[path] = value
        }
    }
}
